### PR TITLE
Add GET endpoint for subconversations

### PR DESCRIPTION
**Pull request by smatting**
This PR:

- refactors MLS functions: use `ConvOrSub` instead of plain conversations
- introduces a GET endpoint for subconversations

Tracked by https://wearezeta.atlassian.net/browse/FS-901.

## Checklist

 - [x] Add a test for `initialGroupId`
 - [ ] Add Stefan M. as a co-author in the merge commit
 - [x] Address the TODO in `Galley.API.Federation` ("TODO: remove this before the rebase")
 - [x] Write golden tests for the `SubConvId` and `PublicSubConversation` types
 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: libs/wire-api-federation/src/Wire/API/Federation/API/Galley.hs [hunk 1]
```diff
@@ -250,18 +251,31 @@ data RemoteMLSMessageResponse
   deriving stock (Eq, Show, Generic)
   deriving (ToJSON, FromJSON) via (CustomEncoded RemoteMLSMessageResponse)
 
-data MessageSendRequest = MessageSendRequest
+data ProteusMessageSendRequest = ProteusMessageSendRequest
```
Comment (elland) [inline libs/wire-api-federation/src/Wire/API/Federation/API/Galley.hs:254]: For next time: maybe worth doing this as laying ground for the change, on a separate PR, to simplify review.

Comment (mdimjasevic) [reply libs/wire-api-federation/src/Wire/API/Federation/API/Galley.hs:254]: That's a good point! Thanks! I'll make this into a separate PR to `develop` and then merge that back into the MLS branch; otherwise we'll be facing conflicts with this in every MLS PR.

Comment (mdimjasevic) [reply libs/wire-api-federation/src/Wire/API/Federation/API/Galley.hs]: This has been done in PR https://github.com/wireapp/wire-server/pull/2925.

**Comment by mdimjasevic** [comment]
@smatting , my apologies, but GitHub somehow doesn't realise anymore that you're a co-author of this PR. I suppose this happened when rebasing on top of `mls` instead of `develop`. Let's not forget to fix that in the merge commit message.

File: libs/wire-api/test/unit/Test/Wire/API/MLS/SubConversation.hs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    genIds :: Gen (ConvId, SubConvId, SubConvId)
+    genIds = do
+      s1 <- arbitrary
+      (,,) <$> arbitrary <*> pure s1 <*> arbitrary `suchThat` (/= s1)
+
+injectiveInitialGroupId :: (ConvId, SubConvId, SubConvId) -> Property
+injectiveInitialGroupId (cnv, scnv1, scnv2) = do
+  let domain = Domain "group.example.com"
+      lcnv = toLocalUnsafe domain cnv
+  initialGroupId lcnv scnv1 =/= initialGroupId lcnv scnv2
```
Comment (elland) [inline libs/wire-api/test/unit/Test/Wire/API/MLS/SubConversation.hs:46]: This tests that they'll (probably) never overlap, but we could also test that the following also holds true.

```
initialGroupId lcnv scnv1 === initialGroupId lcnv scnv1
```

File: services/galley/src/Galley/Cassandra/Queries.hs [hunk 3]
```diff
@@ -313,11 +314,31 @@ deleteUserConv = "delete from user where user = ? and conv = ?"
...
+selectSubConversation = "SELECT cipher_suite, epoch, group_id FROM subconversation WHERE conv_id = ? and subconv_id = ?"
+
+insertSubConversation :: PrepQuery W (ConvId, SubConvId, CipherSuiteTag, Epoch, GroupId, Maybe OpaquePublicGroupState) ()
+insertSubConversation = "INSERT INTO subconversation (conv_id, subconv_id, cipher_suite, epoch, group_id, public_group_state) VALUES (?, ?, ?, ?, ?, ?)"
+
+updateSubConvPublicGroupState :: PrepQuery W (ConvId, SubConvId, Maybe OpaquePublicGroupState) ()
+updateSubConvPublicGroupState = "INSERT INTO subconversation (conv_id, subconv_id, public_group_state) VALUES (?, ?, ?)"
+
+selectSubConvPublicGroupState :: PrepQuery R (ConvId, SubConvId) (Identity (Maybe OpaquePublicGroupState))
+selectSubConvPublicGroupState = "SELECT public_group_state FROM subconversation WHERE conv_id = ? AND subconv_id = ?"
```
Comment (mdimjasevic) [inline services/galley/src/Galley/Cassandra/Queries.hs:335]: These don't really belong to this PR, but I won't insist on removing them from the PR as they'll be needed in the next PR anyway.

**Review by mdimjasevic — APPROVED** [review]

File: libs/wire-api/test/unit/Test/Wire/API/MLS/SubConversation.hs (hunk 2 above)
Comment (elland) [reply libs/wire-api/test/unit/Test/Wire/API/MLS/SubConversation.hs:46]: Fair enough 👍
